File: src/converter/plugins/convertTypeQuery.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";

export const convertTypeQuery = createSimplePlugin((node, context, render) => {
    if (!ts.isTypeQueryNode(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    checkCoverageService?.deepCover(node)

    const resolvedType = typeScriptService?.resolveType(node)

    if (!resolvedType || ts.isTypeQueryNode(resolvedType)) {
        return `Any /* ${typeScriptService?.printNode(node)} */`;
    }

    return render(resolvedType)
})

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertTypeQuery.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import typescript.Node
import typescript.isTypeQueryNode

val convertTypeQuery = createSimplePlugin plugin@{ node: Node, context, render ->
    if (!isTypeQueryNode(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    checkCoverageService?.deepCover(node)

    val resolvedType = typeScriptService?.resolveType(node)

    if (resolvedType == null || isTypeQueryNode(resolvedType)) {
        return@plugin "Any /* ${typeScriptService?.printNode(node)} */"
    }

    render(resolvedType)
}
